### PR TITLE
#1089: IDE2 falls back to new sketch if opening failed.

### LINKED ISSUE #1089: Open Example, Close, Open again - appears to open copy in %temp% directory
### Describe the problem

With the new nightly builds, it looks, like the code that remembers what sketch window that was last closed and tries to reopen it when you restart, it then looks like it reopens the example sketch, which is good.

However what it appears to remember is the directory within your temp directory

Now suppose in between the time you closed the Arduino IDE and reopened the IDE, your folder within temp is no longer there.
Could be you run a disk cleanup, or maybe your rebooted and your temp is on a Ram drive or ...

Now your Arduino IDE comes up in a strange state, something like:
image

Note: it does not show the example sketch.  Also in this case I had open the Sketchbook pane and it showed empty, but Files->Sketchbook does not show empty.

### To reproduce

Open an example Sketch.  In this case I opened MTP_Teensy->SD_MTP_debug  but I don't think that maters. 

Close the sketch: Ignore the error message (say no) about Save/Don't Save  #1026 

Clear out some of the folders within %temp% - like .arduinoIDE-unsaved*

Reopen the Arduino IDE

### Expected behavior

I would expect that it would reopen the example sketch.  Not the copy within the temp directory. 

And/Or - would detect that the folder was not found and maybe inform the user of it and recover.  Could either again go to the Example Sketch,  Could bring up empty sketch or ???

Again would be easier for user to recover from this if the hiccup if like IDE 1.x you update the Recent sketch list, to 
include the examples:  #935 

### Arduino IDE version

Daily build: 20220621

### Operating system

Windows

### Operating system version

Windows 10

### Additional context

Note: This is sort of a follow on and extension of my recent comments in #39 

In particular, have the example open, go away for a couple of hours with the sketch open.  
Have something like Norton running in background, which cleans up files.  Now it clears out the files within the 
.arduino... file in temp as the folder is in use so it can not delete it... 

You then come back to an empty IDE (no file shown) and cryptic message
image
Full image of window in 39... 

You then close the window and reopen and end up with similar to window I show in this issue,  except it knows that it could not open the sketch...
image

### Issue checklist

- [X] I searched for previous reports in [the issue tracker](https://github.com/arduino/arduino-ide/issues?q=)
- [X] I verified the problem still occurs when using the latest [nightly build](https://github.com/arduino/arduino-ide#nightly-builds)
- [X] My report contains all necessary details

### PR DESCRIPTION
**Pull request by kittaakos**
### Motivation

IDE2 must automatically open a new sketch (in the temp folder) if the sketch was deleted between the IDE2 stop/start cycle.

Requirements:
 - At least three valid sketches in the sketchbook folder. My bare minimum setup is the following:
```
a.kitta@Akoss-MacBook-Pro Arduino % ls -al
total 16
drwxr-xr-x  6 a.kitta  staff   192 Jul 14 15:03 .
drwx------+ 5 a.kitta  staff   160 Jul  6 15:46 ..
-rw-r--r--@ 1 a.kitta  staff  6148 Jul 14 15:03 .DS_Store
drwxr-xr-x  3 a.kitta  staff    96 Jul  7 16:42 sketch_1
drwxr-xr-x  3 a.kitta  staff    96 Jul  7 16:43 sketch_2
drwxr-xr-x  3 a.kitta  staff    96 Jul  7 16:43 sketch_3
a.kitta@Akoss-MacBook-Pro Arduino % cat sketch_1/sketch_1.ino 
void setup() {}
void loop() {}
```

(`sketch_2` and `sketch_3` have the same content as `sketch_1`).

Preconditions: 
 - Open IDE2, if `sketch_1` is not loaded, open `sketch_1`,
 - Close any other opened sketches
 - When `sketch_1` is the only opened sketch in IDE2 the verification can start.

1: Opening deleted sketch from `File` > `Sketchbook`
 - Note: the file-system watcher might miss file deletion events and cannot update the `Sketchbook` submenu content.
 - Check `File` > `Sketchbook`, `sketch_3` is there,
 - Delete `sketch_3` outside from IDE2,
 - Open `File` > `Sketchbook` > `sketch_3`,
 - See error notification that it's missing.
 - Check `File` > `Sketchbook`, `sketch_3` is not there.

https://user-images.githubusercontent.com/1405703/179183140-c5018ca3-77e8-4785-af61-01674c718677.mp4

----

2. Delete the sketch folder when IDE2 is stopped:
 - Open `sketch_1` in IDE2,
 - Close IDE2,
 - Delete `sketch_1` from the sketchbook,
 - Start IDE2,
 - Without any error notifications, IDE2 opens a new fallback sketch

https://user-images.githubusercontent.com/1405703/179183207-02ca16e1-1741-477e-8219-e2a1e23dd6bc.mp4

----
3. Delete the main sketch file when IDE2 is stopped:
 - Open `sketch_2` in IDE2,
 - Close IDE2,
 - Delete the main sketch file `sketch_2/sketch_2.ino`,
 - Start the IDE2,
 - IDE2 loads, expect a reload,
 - New sketck is created and opened but no error notification is shown by IDE2.

https://user-images.githubusercontent.com/1405703/179183362-2521a55a-675d-46f9-9a3c-78301cf26523.mp4

### Other information

Closes #1089

Breaking: _Optimize for Debugging_ now has its preference: `arduino.compile.optimizeForDebug`. The default value is `false`. If you had _Optimize for Debugging_ enabled before this IDE2 version, you must click on _Optimized for Debugging_ again.

Screen Shot 2022-07-08 at 11 05 51

Breaking: Changed `arduino.cloud.sketchSyncEnpoint` to `arduino.cloud.sketchSyncEndpoint` (#1067). If you had a custom sketch sync endpoint defined, you must redefine it again with the version of the IDE2.

IDE2 uses the default spinner from Theia:

https://user-images.githubusercontent.com/1405703/178996866-ce705e24-7245-4cee-859f-5ce717ded231.mp4

### Reviewer checklist

* [ ] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [ ] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: arduino-ide-extension/src/node/sketches-service-impl.ts [hunk 1]
```diff
@@ -589,6 +584,14 @@ interface SketchWithDetails extends Sketch {
   readonly mtimeMs: number;
 }
 
+function isNotFoundError(err: unknown): err is ServiceError {
+  return (
+    ServiceError.is(err) &&
+    err.code === 5 &&
+    err.message.toLocaleLowerCase('en-US').includes('not_found')
```
Comment (kittaakos) [inline arduino-ide-extension/src/node/sketches-service-impl.ts:591]: Probably superfluous: https://grpc.github.io/grpc/core/md_doc_statuscodes.html.

File: arduino-ide-extension/src/common/protocol/sketches-service.ts [hunk 2]
```diff
@@ -1,5 +1,21 @@
...
+export namespace SketchesError {
+  export const Codes = {
+    NotFound: 5001,
+  };
+  export const NotFound = ApplicationError.declare(
+    Codes.NotFound,
+    (message: string, uri: string) => {
+      return {
+        message,
+        data: { uri },
```
Comment (kittaakos) [inline arduino-ide-extension/src/common/protocol/sketches-service.ts:13]: Consider including the file-system path? Would be obsolete with the next Theia version.

File: arduino-ide-extension/src/common/protocol/sketches-service-client-impl.ts [hunk 3]
```diff
@@ -110,9 +117,12 @@ export class SketchesServiceClientImpl
...
     });
-    this.loadCurrentSketch().then((currentSketch) =>
-      this._currentSketch.resolve(currentSketch)
-    );
+    setTimeout(async () => {
+      const currentSketch = await this.loadCurrentSketch();
+      this._currentSketch = currentSketch;
+      this.currentSketchDidChangeEmitter.fire(this._currentSketch);
+      this.currentSketchLoaded.resolve(this._currentSketch);
+    }, 1_000);
```
Comment (kittaakos) [inline arduino-ide-extension/src/common/protocol/sketches-service-client-impl.ts:125]: on app `ready`?

File: arduino-ide-extension/src/common/protocol/sketches-service.ts (hunk 2 above)
Comment (kittaakos) [reply arduino-ide-extension/src/common/protocol/sketches-service.ts:13]: 350c04b0bc28bc57acb12d65c6b12cff1777ccea + https://github.com/arduino/arduino-cli/issues/1797

File: arduino-ide-extension/src/common/protocol/sketches-service-client-impl.ts (hunk 3 above)
Comment (kittaakos) [reply arduino-ide-extension/src/common/protocol/sketches-service-client-impl.ts]: ee17fd4

**Review by per1234 — CHANGES_REQUESTED** [review]
There are two distinct reasons the IDE might attempt to open a sketch:

- Explicit action by user (e.g., selecting a sketch from **File > Sketchbook**)
- Restore state from previous session (https://github.com/arduino/arduino-ide/pull/780)

The IDE must gracefully handle a failure of either operation (https://github.com/arduino/arduino-ide/issues/1089). However, I feel the need to communicate about the failure to the user differs between them.

If an operation triggered explicitly by the user fails, it is important to communicate this to the user. So I think showing a notification when the sketch opening fails is necessary in this case.

The situation is different when it comes to automatically opening sketches from the previous IDE session. This is only a convenience feature. If it is able to open the sketches, great. But if it is not able to open the sketches, I don't want to hear about it. I might have intentionally deleted that sketch. It might have been a new sketch that was staged in a temporary folder the OS cleaned up. In the latter case, the temporary path makes the error message especially cryptic:
image

So I suggest that no error notification be shown when the sketch opening fails during the restore operation (logging it is fine of course).

**Comment by kittaakos** [comment]
Thank you for the review!

> So I suggest that no error notification be shown when the sketch opening fails during the restore operation (logging it is fine of course).

Sure 👍 

IDE2 (and the CLI) might fail to load the sketch due to two reasons:
 - the sketch folder was deleted,
 - the sketch is not valid (the main file was deleted, does not match the folder name, .etc)

As a user, do you expect that IDE2 will omit the error notification in both cases?

**Comment by per1234** [comment]
>do you expect that IDE2 will omit the error notification in both cases?

In the case of a "restore" operation, yes.

**Comment by kittaakos** [comment]
I have updated the PR not to show the error notification when restoring the previously opened sketch has failed (due to a missing or invalid sketch). I updated the screencasts too to show the expected IDE2 behavior. Please review.

**Review by per1234 — APPROVED** [review]
It works perfectly. Excellent work Akos!

File: arduino-ide-extension/src/browser/contributions/first-start-installer.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            (await this.boardsService.install({ item: avrPackage }));
+          !!builtInLibrary &&
+            (await this.libraryService.install({
+              item: builtInLibrary,
+              installDependencies: true,
+            }));
+        }
+      });
+  }
+}
```
Comment (AlbyIanna) [inline arduino-ide-extension/src/browser/contributions/first-start-installer.ts:45]: Make sure to remove this as we already put this code [into a separate contribution](https://github.com/arduino/arduino-ide/commit/73835eced3bda7657308bf582bd5e5690d73879a#diff-8e1a8da61c276800216f425dc94832341b6b2b17639b36e837399fb13102b128).

**Review by AlbyIanna — APPROVED** [review]
LGTM.

I left some comments, but nothing important.

File: arduino-ide-extension/src/browser/boards/boards-service-provider.ts [hunk 5]
```diff
@@ -39,6 +39,9 @@ export class BoardsServiceProvider implements FrontendApplicationContribution {
   @inject(NotificationCenter)
   protected notificationCenter: NotificationCenter;
 
+  @inject(FrontendApplicationStateService)
+  private readonly appStateService: FrontendApplicationStateService;
+
```
[the remark this answers is not in the document]
Comment (kittaakos) [reply arduino-ide-extension/src/browser/boards/boards-service-provider.ts:44]: Good idea, but I do not want to introduce a hard dependency between contributions. Also, this service does not contribute any commands, menu items, .etc.

**Pull request merged**